src/DotNet.Testcontainers/Containers/ResourceReaper.cs

```diff
@@ -0,0 +1,228 @@
+namespace DotNet.Testcontainers.Containers
+{
+  using System;
+  using System.Net;
+  using System.Net.Sockets;
+  using System.Text;
+  using System.Threading;
+  using System.Threading.Tasks;
+  using DotNet.Testcontainers.Builders;
+  using DotNet.Testcontainers.Clients;
+  using DotNet.Testcontainers.Configurations;
+
+  public sealed class ResourceReaper : IAsyncDisposable
+  {
+    public const string ResourceReaperSessionLabel = TestcontainersClient.TestcontainersLabel + ".resource-reaper-session";
+
+    private const string RyukImage = "ghcr.io/psanetra/ryuk:2021.12.20";
+
+    private const int RyukPort = 8080;
+
+    private static readonly SemaphoreSlim DefaultLock = new SemaphoreSlim(1, 1);
+
+    private static ResourceReaper defaultInstance;
+
+    private readonly CancellationTokenSource cancellationTokenSource = new CancellationTokenSource();
+
+    private readonly TestcontainersContainer resourceReaperContainer;
+
+    private bool disposed;
+
+    private ResourceReaper(Guid sessionId)
+    {
+      this.resourceReaperContainer = new TestcontainersBuilder<TestcontainersContainer>()
+        .WithName($"testcontainers-ryuk-{sessionId:D}")
+        .WithImage(RyukImage)
+        .WithAutoRemove(true)
+        .WithCleanUp(false)
+        .WithExposedPort(RyukPort)
+        .WithPortBinding(0, RyukPort)
+        .WithBindMount("/var/run/docker.sock", "/var/run/docker.sock", AccessMode.ReadOnly)
+        .Build();
+
+      this.SessionId = sessionId;
+    }
+
+    /// <summary>
+    /// Gets the default <see cref="ResourceReaper" /> session id.
+    /// </summary>
+    /// <remarks>
+    /// The default <see cref="ResourceReaper" /> will start either on <see cref="GetAndStartDefaultAsync(System.Threading.CancellationToken)" />
+    /// or if a <see cref="ITestcontainersContainer" /> is configured with <see cref="ITestcontainersBuilder{TDockerContainer}.WithCleanUp" />.
+    /// </remarks>
+    public static Guid DefaultSessionId { get; }
+      = Guid.NewGuid();
+
+    /// <summary>
+    /// Gets the <see cref="ResourceReaper" /> session id.
+    /// </summary>
+    public Guid SessionId { get; }
+
+    /// <summary>
+    /// Starts and returns the default <see cref="ResourceReaper" /> instance.
+    /// </summary>
+    /// <param name="ct">Cancellation token.</param>
+    /// <returns>Task that completes when the <see cref="ResourceReaper" /> has been started.</returns>
+    public static async Task<ResourceReaper> GetAndStartDefaultAsync(CancellationToken ct = default)
+    {
+      // TODO: Use the State property?
```

Comment:
What state property do you mean and for what purpose?

Reply:
`IRunningDockerContainer.State` aka `this.resourceReaperContainer.State`.

Reply:
I removed the todo. I think it's not necessary.